jerry-core/mem/mem-context.h

```diff
@@ -0,0 +1,119 @@
+/* Copyright 2016 Samsung Electronics Co., Ltd.
+ * Copyright 2016 University of Szeged.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Memory context for JerryScript
+ */
+#ifndef MEM_CONTEXT_H
+#define MEM_CONTEXT_H
+
+#include "jrt.h"
+#include "mem-allocator.h"
+#include "mem-config.h"
+
+/** \addtogroup mem Memory allocation
+ * @{
+ *
+ * \addtogroup heap Context
+ * @{
+ */
+
+/**
+ *  Free region node
+ */
+typedef struct
+{
+  uint32_t next_offset; /* Offset of next region in list */
+  uint32_t size; /* Size of region */
+} mem_heap_free_t;
+
+/**
+ * Node for free chunk list
+ */
+typedef struct mem_pools_chunk
+{
+  struct mem_pools_chunk *next_p; /* pointer to next pool chunk */
+} mem_pools_chunk_t;
+
+/**
+ * Calculate heap area size, leaving space for a pointer to the free list
+ */
+#define MEM_HEAP_AREA_SIZE (MEM_HEAP_SIZE - MEM_ALIGNMENT)
+
+/**
+ * Heap structure
+ */
+typedef struct
+{
+  mem_heap_free_t first; /**< first node in free region list */
+  uint8_t area[MEM_HEAP_AREA_SIZE]; /**< heap area */
+} mem_heap_t;
+
+/**
+ * JerryScript context
+ *
+ * The purpose of this header is storing
+ * all global variables for Jerry
+ */
+typedef struct
+{
+  /**
+   * Memory manager part.
+   */
+  size_t mem_heap_allocated_size; /**< size of allocated regions */
+  size_t mem_heap_limit; /**< current limit of heap usage, that is upon being reached,
+                          *   causes call of "try give memory back" callbacks */
+  mem_heap_free_t *mem_heap_list_skip_p; /**< This is used to speed up deallocation. */
```

Comment:
The "context" is quite general for me. It should contain everything except heap and lcache (those two are big arrays). My plan for the layout when context is dynamically allocated:

    [ context ][ heap ][ lcache ]

The context pointer points to the start of the heap. The context members are accessed by 

   ((mem_context_t*) heap_ptr)[-1].member

Even on Thumb2 ARM +- 255 offset is supported, so we can access the fields of the context easily (provided the size of the context is < 255, but I do hope it will be). We should not have too many global variables.

Similar to the previous one, I am open to suggestions here.